src/Nethermind/Nethermind.Consensus/Processing/CensorshipDetector/CensorshipDetector.cs

```diff
@@ -0,0 +1,228 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Nethermind.Core;
+using Nethermind.Core.Caching;
+using Nethermind.Core.Crypto;
+using Nethermind.Crypto;
+using Nethermind.Int256;
+using Nethermind.Logging;
+using Nethermind.TxPool;
+
+namespace Nethermind.Consensus.Processing.CensorshipDetector;
+
+public class CensorshipDetector : IDisposable
+{
+    private readonly ITxPool _txPool;
+    private readonly IComparer<Transaction> _comparer;
+    private readonly IBlockProcessor _blockProcessor;
+    private readonly ILogger _logger;
+    private readonly ICensorshipDetectorConfig _censorshipDetectorConfig;
+    private readonly Dictionary<AddressAsKey, Transaction?> _poolAddressCensorshipBestTx = [];
+    private readonly LruCache<BlockNumberHash, BlockCensorshipInfo> _potentiallyCensoredBlocks;
+    private readonly WrapAroundArray<BlockNumberHash> _censoredBlocks;
+    private const int _cacheSize = 64;
+
+    public CensorshipDetector(
+        ITxPool txPool,
+        IComparer<Transaction> comparer,
+        IBlockProcessor blockProcessor,
+        ILogManager logManager,
+        ICensorshipDetectorConfig censorshipDetectorConfig)
+    {
+        _txPool = txPool;
+        _comparer = comparer;
+        _blockProcessor = blockProcessor;
+        _logger = logManager?.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
+        _censorshipDetectorConfig = censorshipDetectorConfig;
+
+        if (_censorshipDetectorConfig.AddressesForCensorshipDetection is not null)
+        {
+            foreach (string hexString in _censorshipDetectorConfig.AddressesForCensorshipDetection)
+            {
+                if (Address.TryParse(hexString, out Address address))
+                {
+                    _poolAddressCensorshipBestTx[address!] = null;
+                }
+            }
+        }
+
+        _potentiallyCensoredBlocks = new(_cacheSize, _cacheSize, "potentiallyCensoredBlocks");
+        _censoredBlocks = new(_cacheSize);
+        _blockProcessor.BlockProcessing += OnBlockProcessing;
+    }
+
+    private void OnBlockProcessing(object? sender, BlockEventArgs e)
+    {
+        UInt256 baseFee = e.Block.BaseFeePerGas;
+
+        IEnumerable<Transaction> poolBestTransactions = _txPool.GetBestTxOfEachSender();
```

Comment:
One thing, maybe instead of getting all from pool, we should loop and get best only for tracked addresses, we will pull only that number of transactions from pool instead of pulling up to 2k transactions here.

So instead of having `GetBestTxOfEachSender()` we could have `GetBestTransaction(Address sender)`.

Also we need to check if this transaction is includable?

Reply:
Our tracked addresses are in all TxPool txs as `To`, not senders. So we will still need to iterate

Reply:
ok ignore then, bad idea